tests/bot/cogs/test_token_remover.py

```diff
@@ -1,131 +1,279 @@
 
-        for content, censored_token in cases:
-            with self.subTest(content=content, censored_token=censored_token):
-                self.msg.content = content
-                coroutine = self.cog.on_message(self.msg)
-                with self.assertLogs(logger='bot.cogs.token_remover', level=logging.DEBUG) as cm:
-                    self.assertIsNone(asyncio.run(coroutine))  # no return value
-
-                [line] = cm.output
-                log_message = (
-                    "Censored a seemingly valid token sent by "
-                    "lemon (`42`) in #lemonade-stand, "
-                    f"token was `{censored_token}`"
-                )
-                self.assertIn(log_message, line)
-
-                self.msg.delete.assert_called_once_with()
-                self.msg.channel.send.assert_called_once_with(
-                    DELETION_MESSAGE_TEMPLATE.format(mention='@lemon')
-                )
-                self.bot.get_cog.assert_called_with('ModLog')
-                self.msg.author.avatar_url_as.assert_called_once_with(static_format='png')
-
-                mod_log = self.bot.get_cog.return_value
-                mod_log.ignore.assert_called_once_with(Event.message_delete, self.msg.id)
-                mod_log.send_log_message.assert_called_once_with(
-                    icon_url=Icons.token_removed,
-                    colour=Colour(Colours.soft_red),
-                    title="Token removed!",
-                    text=log_message,
-                    thumbnail='picture-lemon.png',
-                    channel_id=Channels.mod_alerts
-                )
-
-
-class TokenRemoverSetupTests(unittest.TestCase):
-    """Tests setup of the `TokenRemover` cog."""
-
-    def test_setup(self):
-        """Setup of the extension should call add_cog."""
+    @autospec("bot.cogs.token_remover", "LOG_MESSAGE")
+    async def test_format_log_message(self, log_message):
+        """Should correctly format the log message with info from the message and token."""
+        log_message.format.return_value = "Howdy"
+        return_value = TokenRemover.format_log_message(self.msg, "MTIz.DN9R_A.xyz")
+
+        self.assertEqual(return_value, log_message.format.return_value)
+        log_message.format.assert_called_once_with(
+            author=self.msg.author,
+            author_id=self.msg.author.id,
+            channel=self.msg.channel.mention,
+            user_id="MTIz",
+            timestamp="DN9R_A",
+            hmac="xxx",
+        )
+
+    @mock.patch.object(TokenRemover, "mod_log", new_callable=mock.PropertyMock)
+    @autospec("bot.cogs.token_remover", "log")
+    @autospec(TokenRemover, "delete_message", "format_log_message")
+    async def test_take_action(self, delete_message, format_log_message, logger, mod_log_property):
+        """Should delete the message and send a mod log."""
+        cog = TokenRemover(self.bot)
+        mod_log = mock.create_autospec(ModLog, spec_set=True, instance=True)
```

Comment:
Could `mod_log` not just use the `@autospec` decorator since it allows any valid `**kwargs` for `patch.object()`?

E.g. `@autospec(TokenRemover, "mod_log", new_callable=mock.PropertyMock)` instead of having `@mock.patch.object(TokenRemover, "mod_log", new_callable=mock.PropertyMock)` and `mod_log = mock.create_autospec(ModLog, spec_set=True, instance=True)`? It would behave the same AFAICT, but it seems worthwhile to remove a line of boilerplate.

Note that I don't personally have much experience w/ making use of mocks to this degree in my own unit tests, so the above is more of a question as it is a suggestion.

Reply:
`new_callable` cannot be used if autospeccing. I could set `autospec=False` but at that point I felt like I should use the normal unittest decorators.